cmd/kubeadm/app/api/types.go

```diff
@@ -16,89 +16,54 @@ limitations under the License.
 
 package api
 
-import (
-	"net"
-)
+import "k8s.io/kubernetes/pkg/api/unversioned"
 
-// KubeadmConfig TODO add description
-// TODO(phase1+) @krousey: Please don't embed structs. It obfuscates the source of the fields and doesn't really buy you anything.
-type KubeadmConfig struct {
-	InitFlags
-	JoinFlags
-	Secrets struct {
-		GivenToken  string // dot-separated `<TokenID>.<Token>` set by the user
-		TokenID     string // optional on master side, will be generated if not specified
-		Token       []byte // optional on master side, will be generated if not specified
-		BearerToken string // set based on Token
-	}
-	EnvParams map[string]string // TODO(phase2) this is likely to be come componentconfig
-}
-
-// TODO(phase2) should we add validation functions for these structs?
+type MasterConfiguration struct {
+	unversioned.TypeMeta
 
-// TODO(phase1+) refactor token handling
-// - https://github.com/kubernetes/kubernetes/pull/33262/files#r80333662
-// - https://github.com/kubernetes/kubernetes/pull/33262/files#r80336374
-// - https://github.com/kubernetes/kubernetes/pull/33262/files#r80333982
+	Secrets           Secrets
+	API               API
+	Etcd              Etcd
+	Networking        Networking
+	KubernetesVersion string
+	CloudProvider     string
+}
 
-// InitFlags holds values for "kubeadm init" command flags.
-type InitFlags struct {
-	API struct {
-		AdvertiseAddrs   []net.IP
-		ExternalDNSNames []string
-		Etcd             struct {
-			ExternalEndpoints []string
-			ExternalCAFile    string
-			ExternalCertFile  string
-			ExternalKeyFile   string
-		}
-	}
-	Services struct {
-		CIDR      net.IPNet
-		DNSDomain string
-	}
-	PodNetwork struct {
-		CIDR net.IPNet
-	}
-	Versions struct {
-		Kubernetes string
-	}
-	CloudProvider string
+type API struct {
+	AdvertiseAddresses []string
+	ExternalDNSNames   []string
 }
 
-const (
-	DefaultServiceDNSDomain   = "cluster.local"
-	DefaultServicesCIDRString = "100.64.0.0/12" // Carrier-grade NAT range (RFC 6598)
-	DefaultKubernetesVersion  = "v1.4.0"
-)
+type Networking struct {
+	ServiceSubnet string
```

Comment:
I'm getting this ready to become an API group which won't support that type. I suppose I can transistion it in a follow up.